### ISSUE TITLE
valid-jsdoc rule doesn't work on @param {String} options.X-My-Header custom header

### ISSUE THREAD

**Issue by nfriedly** [eslint/eslint#5612]
**What version of ESLint are you using?**
v2.2.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

Full config is at https://github.com/watson-developer-cloud/speech-javascript-sdk/blob/51dd4a041c8e74051ebb87b2630f040b6f0f19a9/.eslintrc.js#L205 but here's the relevant bit.

```js
        'valid-jsdoc': [1, {
          'requireReturn': false,
          'requireParamDescription': false,
          'requireReturnDescription': false
        }],
```
 I also tested and confirmed that I get the same issue with just `'valid-jsdoc': 1`.

**What did you do? Please include the actual source code causing the issue.**

This is one example of my (working) JSDoc that's failing eslint:

` * @param {Number} [options.X-WDC-PL-OPT-OUT=0] set to 1 to opt-out of allowing Watson to use this request to improve it's services`

You can see it in context at https://github.com/watson-developer-cloud/speech-javascript-sdk/blob/9b683539218a3f060116aa65cfa709380a3ae98c/text-to-speech/synthesize.js#L34

I also tried writing it as

` * @param {Number} [options['X-WDC-PL-OPT-OUT']=0] set to 1 to opt-out of allowing Watson to use this request to improve it's services`

That passes eslint but causes JSDoc to explode. (I filed a bug about that at https://github.com/jsdoc3/jsdoc/issues/1178 - but I still think the first behavior is an eslint bug: a rule named "valid-jsdoc" really shouldn't enforce invalid JSDoc.)

**What did you expect to happen?**

I expect my working JSDoc to pass the valid-jsdoc rule, or for there to at least be some format that works for both eslint and JSDoc

**What actually happened? Please include the actual, raw output from ESLint.**

That line gives me 

`26:1  warning  Missing JSDoc parameter type for 'Number'  valid-jsdoc`

Here's the full output:

```
~/watson-speech (master) $ npm run lint

> watson-speech@0.14.0 lint /Users/nfriedly/watson-speech
> eslint .


/Users/nfriedly/watson-speech/speech-to-text/format-stream.js
  113:7  warning  Unexpected 'todo' comment  no-warning-comments

/Users/nfriedly/watson-speech/speech-to-text/media-element-audio-stream.js
  68:7  warning  Unexpected 'todo' comment  no-warning-comments

/Users/nfriedly/watson-speech/speech-to-text/recognize-stream.js
   34:1   warning  Missing JSDoc parameter type for 'String'  valid-jsdoc
   34:1   warning  Missing JSDoc parameter type for 'Number'  valid-jsdoc
  146:13  warning  Unexpected 'todo' comment                  no-warning-comments
  149:9   warning  Unexpected 'todo' comment                  no-warning-comments
  165:3   warning  Unexpected 'todo' comment                  no-warning-comments
  379:1   warning  Unexpected 'todo' comment                  no-warning-comments

/Users/nfriedly/watson-speech/speech-to-text/timing-stream.js
  37:32  warning  Unexpected 'todo' comment  no-warning-comments

/Users/nfriedly/watson-speech/text-to-speech/synthesize.js
  26:1  warning  Missing JSDoc parameter type for 'Number'  valid-jsdoc

✖ 10 problems (0 errors, 10 warnings)
```

(Ignore the todo's - I want those to nag me until they're fixed.)

**Comment by ilyavolodin** [comment]
ESLint uses Doctrine as a parser for jsdoc, so I think you might've filed an issue in the wrong repository (jsdoc). I think the rule is complaining because you are trying to describe properties of an object, where your function accepts only one object. I'm actually not sure how this is supposed to work. @gyandeeps any ideas about the right way to do jsdoc comments for object properties of an argument?

**Comment by BYK** [comment]
I think he needs to define `options` using `@type` first and then simply say `{OptionsObject} [options]`.

**Comment by nfriedly** [comment]
OK, to give some expanded context, here's what the full JSDoc for that method looks like:

```js
/**
 * Synthesize and play the supplied text over the computers speakers.
 *
 * Creates and returns a HTML5 `<audio>` element
 *
 * @param options
 * @param {String} options.token auth token
 * @param {String} options.text text to speak
 * @param {String} [options.voice=en-US_MichaelVoice] what voice to use - call getVoices() for a complete list.
 * @param {Number} [options.X-WDC-PL-OPT-OUT=0] set to 1 to opt-out of allowing Watson to use this request to improve it's services
 * @param {Boolean} [options.autoPlay=true] automatically play the audio
 * @returns {Audio}
 * @see module:watson-speech/text-to-speech/get-voices
 */
```

And JSDoc produces a table from that with each property listed: http://watson-developer-cloud.github.io/speech-javascript-sdk/v0.14.0/module-watson-speech_text-to-speech_synthesize.html

The whole codebase uses this pattern, and except for hyphenated object keys, everything works swimmingly. But for hyphenated keys, JSDoc and ESLint/Doctrine seem to differ in what they'll accept. 

I'll try it with a `@type` and let you know if that works any better, but it probably won't be until Monday. 

I'll also take a look at Doctrine and probably file a bug there.

**Comment by BYK** [comment]
@nfriedly - thanks for sharing the larger snippet. Now I can definitely see this is an issue. This is most probably due to https://github.com/eslint/doctrine/issues/116

I'll keep this ticket here in ESLint open to track progess.

**Comment by kaicataldo** [comment]
Opened a [PR to doctrine](https://github.com/eslint/espree/pull/265) to attempt to fix this. I've already tested that branch of doctrine with ESLint locally and can make a PR with tests covering this and bumping the doctrine version if that gets merged.

**Comment by kaicataldo** [comment]
Changes merged to doctrine. Will finalize this PR when the next doctrine version is released.

**Comment by kaicataldo** [comment]
PR opened with the bumped version of Doctrine and some tests covering this case

**Comment by nfriedly** [comment]
BTW, I just wanted to mention that this is now working for me. Thanks everyone!

**Comment by kaicataldo** [comment]
@nfriedly Thanks for the kind words 😄
